### DISCUSSION TITLE
How to debug betaFPV micro 2.4Ghz/1000mW module

### DISCUSSION THREAD

**Discussion by mha1** [ExpressLRS/ExpressLRS#2034] [Q&A]
Hi,

how do build firmware to use DBGLN() and how to make a connection to the module to see the DBGLN() serial output? I guess defining DEBUG_LOG in the user defines (partially?) answers the first part.

Can someone please get me started?

Thanks - Michael

**Comment by pkendall64** [comment]
You would also need to edit the /hardware.html page on the web-ui and change the backpack to disabled and set the backpack/debug pins to 1 for RX and 3 for TX then it will send debug info out the USB port.

**Comment by mha1** [comment]
@pkendall64 Thank you very much. I can't see any defaults in `src/html/hardware.html`. Do you mean to edit the  `src/TX/<module>.json`? Like this to disable backpack, RX pin 3, TX pin 1?
```
    "use_backpack": true,
    "debug_backpack_baud": 460800,
    "debug_backpack_rx": 3,
    "debug_backpack_tx": 1,
```
to
```
    "use_backpack": false,
    "debug_backpack_baud": 460800,
    "debug_backpack_rx": 3,
    "debug_backpack_tx": 1,
```

**Reply by pkendall64** [reply]
You should see the values in the hardware.html, but since this is the BetaFPV 1W module, you don't need those changes, just put the switches on the back of the module into ESP32 flash mode and that means that pins 1 and 3 are connected to the USB port rather than the ESP8285 backpack chip.

**Comment by mha1** [comment]
Ok, will try.

But back to the html. We are talking about src/html/hardware.html, aren't we? I found the backpack section but this only defines the HTML layout:
```
<tr><td colspan='2'><b>Backpack / Logging</td></tr>
<tr><td></td><td>Enable Backpack</td><td><input size='3' id='use_backpack' name='use_backpack' type='checkbox'/></td><td>If a TX backpack is connected</td></tr>
<tr><td></td><td>Baud Rate</td><td><input size='10' id='debug_backpack_baud' name='debug_backpack_baud' type='text'/></td><td>Baud rate used to communicate to the backpack (normally 460800)</td></tr>
<tr><td></td><td>RX pin<img class="icon-input"/></td><td><input size='3' id='debug_backpack_rx' name='debug_backpack_rx' type='text'/></td><td>Connected to TX pin on backpack</td></tr>
<tr><td></td><td>TX pin<img class="icon-output"/></td><td><input size='3' id='debug_backpack_tx' name='debug_backpack_tx' type='text'/></td><td>Connected to RX pin on backpack</td></tr>
<tr><td></td><td>BOOT pin<img class="icon-output"/></td><td><input size='3' id='backpack_boot' name='backpack_boot' type='text'/></td><td>Pin connected to GPIO0 pin on backpack ESP8285, allows passthrough flashing</td></tr>
<tr><td></td><td>EN pin<img class="icon-output"/></td><td><input size='3' id='backpack_en' name='backpack_en' type='text'/></td><td>Pin connected to EN pin on backpack ESP8285, allows passthrough flashing</td></tr>
<tr><td></td><td>Passthrough baud</td><td><input size='7' id='passthrough_baud' name='passthrough_baud' type='text'/></td><td>Baud rate to flash the backpack ESP8285 (default is to use the baud rate above)</td></tr>
```
Where can I find the defaults to be able to change them? 

And a completly stupid question: where and how would I see this html page?

**Reply by pkendall64** [reply]
You need to let the device go into wifi mode then go to http://elrs_tx.local/hardware.html then it will display the page filled with the current hardware layout.json file, which you can edit and save. The idea is that for we can upload a bare firmware (no hardware definition) and a user or manufacturer can configure it on that page. That way we only need a single binary for each device type.

To change the defaults you would edit the appropriate json file under the `src/hardware` directory in the git repository.

Hope that helps.

**Comment by mha1** [comment]
@pkendall64

Paul, thanks again. Building with -DDEBUG_LOG and setting DIP switches 1 and 2 to on works. Debug output is sent to the USB-C port at 460800 baud.

But I couldn't find the settings in `src/html/hardware.html`, just the html layout. Are you sure the default settings aren't defined in  `src/TX/<module>.json`?

And out of curiosity: how do you access the html page defined in `src/html/hardware.html`?

Michael

**Comment by mha1** [comment]
Oops, race condition. You won.

Thank you very much, got it!
